src/EventStore.Core/Services/RedactionService.cs

```diff
@@ -0,0 +1,261 @@
+
+			if (IsUnsafeFileName(targetChunkFile)) {
+				failReason = SwitchChunkResult.TargetChunkFileNameInvalid;
+				return false;
+			}
+
+			if (IsUnsafeFileName(newChunkFile)) {
+				failReason = SwitchChunkResult.NewChunkFileNameInvalid;
+				return false;
+			}
+
+			int targetChunkNumber;
+			try {
+				targetChunkNumber = _db.Config.FileNamingStrategy.GetIndexFor(targetChunkFile);
+			} catch {
+				failReason = SwitchChunkResult.TargetChunkFileNameInvalid;
+				return false;
+			}
+
+			if (Path.GetExtension(newChunkFile) != NewChunkFileExtension) {
+				failReason = SwitchChunkResult.NewChunkFileNameInvalid;
+				return false;
+			}
+
+			if (!File.Exists(Path.Combine(_db.Config.Path, targetChunkFile))) {
+				failReason = SwitchChunkResult.TargetChunkFileNotFound;
+				return false;
+			}
+
+			var newChunkPath = Path.Combine(_db.Config.Path, newChunkFile);
+			if (!File.Exists(newChunkPath)) {
+				failReason = SwitchChunkResult.NewChunkFileNotFound;
+				return false;
+			}
+
+			TFChunk targetChunk;
+			try {
+				targetChunk = _db.Manager.GetChunk(targetChunkNumber);
+			} catch(ArgumentOutOfRangeException) {
+				failReason = SwitchChunkResult.TargetChunkExcessive;
+				return false;
+			}
+
+			if (Path.GetFileName(targetChunk.FileName) != targetChunkFile) {
+				failReason = SwitchChunkResult.TargetChunkInactive;
+				return false;
+			}
+
+			if (targetChunk.ChunkFooter is not { IsCompleted: true }) {
+				failReason = SwitchChunkResult.TargetChunkNotCompleted;
+				return false;
+			}
+
+			ChunkHeader newChunkHeader;
+			ChunkFooter newChunkFooter;
+			try {
+				using var fs = new FileStream(newChunkPath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
+				try {
+					newChunkHeader = ChunkHeader.FromStream(fs);
+					fs.Seek(-ChunkFooter.Size, SeekOrigin.End);
+					newChunkFooter = ChunkFooter.FromStream(fs);
+				} catch {
+					failReason = SwitchChunkResult.NewChunkHeaderOrFooterInvalid;
+					return false;
+				}
+			} catch {
+				failReason = SwitchChunkResult.NewChunkOpenFailed;
+				return false;
+			}
+
+			if (newChunkHeader.ChunkStartNumber != targetChunk.ChunkHeader.ChunkStartNumber ||
+			    newChunkHeader.ChunkEndNumber != targetChunk.ChunkHeader.ChunkEndNumber) {
+				failReason = SwitchChunkResult.ChunkRangeDoesNotMatch;
+				return false;
+			}
+
+			if (!newChunkFooter.IsCompleted) {
+				failReason = SwitchChunkResult.NewChunkNotCompleted;
+				return false;
+			}
+
+			try {
+				newChunk = TFChunk.FromCompletedFile(
+					filename: newChunkPath,
+					verifyHash: true,
+					unbufferedRead: _db.Config.Unbuffered,
+					initialReaderCount: 1,
+					maxReaderCount: 1,
+					optimizeReadSideCache: false,
+					reduceFileCachePressure: true);
+			} catch (HashValidationException) {
+				failReason = SwitchChunkResult.NewChunkHashInvalid;
+				return false;
+			} catch {
+				failReason = SwitchChunkResult.NewChunkOpenFailed;
+				return false;
+			} finally {
+				newChunk?.Dispose();
```

Comment:
is this intentional?

Reply:
yes, if we move `newChunk.Dispose();` to the try block, and calling Dispose() throws, then the reply reason will be: `SwitchChunkResult.NewChunkOpenFailed`. but right now if Dispose() throws, it'll return `SwitchChunkResult.UnexpectedError` which is more appropriate

Reply:
hmmm for more clarity, i'm moving it out of the finally block, since it's not necessary (as newChunk will not be assigned anyway when exceptions are thrown)

Reply:
agreed, fixed in: https://github.com/EventStore/EventStore/pull/3713/commits/a9533ba7848f78ead50abbf304e3dce504430b53 and added a comment

Reply:
still suspicious to pass a disposed chunk in to switch chunk? 😅

Reply:
hehe, fair enough, i've removed the Dispose() call :)